src/test/java/com/google/graphgeckos/dashboard/api/DashboardControllerTest.java

```diff
@@ -0,0 +1,165 @@
+  public void setUpBuildInfo() {
+    BUILD_INFO.addBuilder(BUILDER_A);
+    BUILD_INFO.addBuilder(BUILDER_B);
+  }
+
+  /**
+   * Given: number of the entries to return = 0, any offset >= 0
+   * Expected behaviour: empty JSON list, Response Status 200 (OK).
+   */
+  @Test
+  public void whenGivenZeroRevisionsOffsetZeroReturnsEmptyJSONArray() throws Exception {
+
+    given(datastoreRepository.getLastRevisionEntries(ZERO_ENTRIES, OFFSET_ZERO))
+      .willReturn(Collections.emptyList());
+
+    ResultActions result = mvc.perform(
+      MockMvcRequestBuilders.get("/builders/number={number}/offset={offset}",
+        ZERO_ENTRIES, OFFSET_ZERO)
+        .accept(MediaType.APPLICATION_JSON));
+
+    String emptyJsonList = "[]";
+    result.andExpect(status().isOk())
+          .andExpect(content().contentType(MediaType.APPLICATION_JSON))
+          .andExpect(content().string(emptyJsonList));
+  }
+
+  /**
+   * Should return the JSON list of length.
+   *
+   * Given: number of entries to return >= 0; offset b, b >= 0.
+   * Expected behaviour: JSON list of length = number
+   * of entries to return, more about offset and number at
+   * {@link com.google.graphgeckos.dashboard.storage.DataRepository::getLastRevision}.
+   * Returned list contains JSON representations of the BuildBotInfo instances.
+   * Information about this JSON representations could be found at
+   * {@link com.google.graphgeckos.dashboard.datatypes.BuildInfo}.
+   */
+  @Test
+  public void whenGivenOneRevisionOffsetZeroReturnsOneBuildInfoJson() throws Exception {
+
+    given(datastoreRepository.getLastRevisionEntries(ONE_ENTRIES, OFFSET_ZERO))
+      .willReturn(Collections.singletonList(BUILD_INFO));
+
+    ResultActions result = mvc.perform(
+        MockMvcRequestBuilders.get("/builders/number={number}/offset={offset}",
+          ONE_ENTRIES, OFFSET_ZERO)
+       .accept(MediaType.APPLICATION_JSON));
+
+    result.andExpect(status().isOk())
+          .andExpect(content().contentType(MediaType.APPLICATION_JSON))
+          .andExpect(jsonPath("$[0].commitHash").value(COMMIT_HASH))
+          .andExpect(jsonPath("$[0].branch").value(BRANCH))
+          .andExpect(jsonPath("$[0].builders[0].name").value(NAME_A))
+          .andExpect(jsonPath("$[0].builders[1].status").value(STATUS_B))
+          .andExpect(jsonPath("$[0].builders[0].logs[1].type").value(LOG_TYPE_2))
+          .andExpect(jsonPath("$[0].builders[0].logs[0].link").value(LOG_LINK_1));
+  }
+
+  /**
+   * Number of entries and offset should be greater or equal to zero, otherwise the IllegalArgumentException is
+   * thrown inside the tested request method. If the described situation occurs, the request method
+   * should set the response status to "Bad request".
+   *
+   * Given: illegal argument value/values.
+   * Expected behaviour: response status 400(Bad request).
+   */
+  @Test
+  public void setsRequestTypeToBadRequestWhenIllegalArgumentExceptionIsThrownByDatastoreRepository() throws Exception {
+
```

Comment:
nit: remove blank line at the start of the code

Reply:
Same as above